Codeplay graphs vendor extension

EwanC opened the pull request:
Make public the specification for `SYCL_EXT_CODEPLAY_GRAPHS`, Codeplay's vendor extension introducing an API for user capture and submission of command-group graphs.

This is the initial version and still subject to change.

Commit:
Codeplay graphs vendor extension

Make public the specification for `SYCL_EXT_CODEPLAY_GRAPHS`, Codeplay's
vendor extension introducing an API for user capture and submission of
command-group graphs.

This is the initial version and still subject to change.

reble commented on graphs/index.md:
```diff
@@ -0,0 +1,706 @@
+#### Single Queue Submit
+
+This extension allows a graph to be recorded from multiple different queues,
+but ony submitted to a single queue. The device associated with a queue is not
```
```suggestion
but only submitted to a single queue. The device associated with a queue is not
```

reble commented on graphs/index.md:
```diff
@@ -0,0 +1,706 @@
+
+// State of a graph
+enum class graph_state {
+  recordable,
```
Naming this state `modifiable` or `editable` might be more generic, and better support future extensions (such as an Explicit API). Or would you consider adding an extra state?

EwanC replied:
We weren't considering adding extra state, so calling `recordable` more generic would indeed to better introducing an explicit API. My worry with `modifiable` or `editable` is that the user can perform a kind of `modify` or `edit`  of the graph when we update it, either through whole-graph update or individual node update in an Explicit API.

Would probably tend towards `constructable`, `buildable`, `composable`, or something along those lines. Will have a think.

EwanC replied:
Decided to make this `modifiable` for the moment to reduce the diff with the oneAPI extension, once they are merged we can always revisit this.

reble commented on graphs/index.md:
```diff
@@ -0,0 +1,706 @@
+
+A `command_graph` object is constructed in the *recording* state and is made
+*executable* by the user invoking `command_graph::finalize()` to create a
+new executable instance of the graph. An *executable* graph cannot transition
```
Transition could be misleading here. My understanding is, you keep the *editable* `command_graph` object alive, and create (multiple) *executable* object(s) from it, correct? Can `finalize()` method be `const`?

EwanC replied:
> My understanding is, you keep the editable command_graph object alive, and create (multiple) executable object(s) from it, correct?

Yup, you understand correctly. *transition* does indeed sound like there is state change happening in-place, rather than creating something new. Will revisit this wording.

> `Can finalize()` method be const?

Not as we've currently implemented it, but conceptually I think this should be possible, and would make it clearer to the user about the point above of not modifying state in-place. Will take a look at refactoring that section of code to make this happen.

EwanC replied:
We tweaked our implementation and indeed this can be `const`, updated to specify that

Commit:
Rename recordable graph state

Rename the `recordable` graph state to `modifiable`, so that
is more generic for when an explicit API is added.

Additionally, make `finalize()` a `const` function.
